### ISSUE TITLE
Input SQL Server - Add "Query Stats"

### ISSUE THREAD

**Issue by Trovalo** [influxdata/telegraf#7789]
## Feature Request

Add data from sys.dm_exec_query_stats, in order to track frequent heavy queries.

### Proposal:
Add a measurement with the data of the DMV sys.dm_exec_query_stats

### Notes:
- due to the high number of rows, the result should be limited to the top N rows (ie TOP 50 rows) by "Total Execution time"
- A sorting order should be defined, "total_elapsed_time" could be a good choice
- of all the column available (total, last, min, max), taking only the "total_*" should suffice, since the other can be calculated

Here is a draft of the query I want to use (not yet with retro compatibility), some columns have been added in SQL 2012, 2014 and 2016.
```SQL
USE [master]

SELECT TOP 50
    'sqlserver_requests' AS [measurement]
    ,REPLACE(@@SERVERNAME,'\',':') AS [sql_instance]
    ,DB_NAME() as [database_name]
    ,qs.[query_hash]
    ,qs.[query_plan_hash]
	,qs.[plan_generation_num]
    --,qs.[statement_start_offset]
    --,qs.[statement_end_offset]
    ,qs.[execution_count]
    ,qs.[total_worker_time]/1000 AS [total_worker_time_ms]
    ,qs.[total_physical_reads]
    ,qs.[total_logical_writes]
    ,qs.[total_logical_reads]
    ,qs.[total_clr_time]/1000 AS [total_clr_time_ms]
    ,qs.[total_elapsed_time]/1000 AS [total_elapsed_time_ms]
    ,qs.[total_rows]
    --,qs.[statement_sql_handle]
    --,qs.[statement_context_id]
    ,qs.[total_dop]
    ,qs.[total_grant_kb]
    ,qs.[total_used_grant_kb]
    ,qs.[total_ideal_grant_kb]
    ,qs.[total_reserved_threads]
    ,qs.[total_used_threads]
    ,qs.[total_columnstore_segment_reads]
    ,qs.[total_columnstore_segment_skips]
    ,qs.[total_spills]
	,QUOTENAME(OBJECT_SCHEMA_NAME(qt.objectid,qt.dbid)) + '.' +  QUOTENAME(OBJECT_NAME(qt.objectid,qt.dbid)) as stmt_object_name
	,(SUBSTRING(qt.text, qs.[statement_start_offset] / 2 + 1,
		(CASE WHEN qs.[statement_end_offset] = -1
				THEN LEN(CONVERT(NVARCHAR(MAX), qt.text)) * 2
				ELSE qs.[statement_end_offset]
		END - qs.[statement_start_offset]) / 2)
	) AS statement_text
	,DB_NAME(qt.[dbid]) stmt_db_name
FROM sys.dm_exec_query_stats as qs
OUTER APPLY sys.dm_exec_sql_text(qs.[sql_handle]) AS qt
ORDER BY
    [total_elapsed_time] DESC
```

**Comment by Trovalo** [comment]
@denzilribeiro any thought about this?

**Comment by denzilribeiro** [comment]
I personally think this can be got on query store for most of SQL 2016 or later. It isn't very easy to visualize any of this. The interval of 10 seconds is WAY to aggressive for this type of aggregation. I personally don't think query level analysis is suited for Telegraf rather more of monitoring metrics.
From a query perspective, am pretty sure many of the columns aren't there on earlier versions, and some are not added from newer versions. This type of collection is much easier if could specify different time frame (once every 5 mins). Definitely would not collect this by default at a 10/15s interval will be way too much data and not enough time to ever aggregate.

**Comment by Trovalo** [comment]
I agree on the gathering frequency, it does not make a lot of sense to gather this every few seconds, but the frequency can be changed (even if that requires a different sqlserver section in the configuration, which personally is something I already use to gather data at different frequencies)

**Comment by powersj** [comment]
@Trovalo, any desire to progress this one along still?

**Comment by Trovalo** [comment]
I forgot about it, The feature itself has been running without issues on my own build for months now but I'm not sure about adding it to the available queries the reasons are:
- This is an advanced analysis based on cumulative data, the result of the query as-is is not useful and might even be misleading (must be "converted" based on gather frequency)
- The query might be pretty heavy (a specific schedule (let's say every 5m) is recommended -> interval override)
- The query may produce a variable amount of data (limited to 100 rows), that might be pretty heavy since also the text is captured

In summary, the query is very useful but also "hard" to manage and the setup to have it working properly is not straightforward at all. 
I honestly won't make it available, as it might cause more harm than good.
Let me know if that makes sense to you

_note: I'm not even sure this branch is updated to my latest version_

**Comment by powersj** [comment]
Thanks for the detailed explanation. I agree that this is probably not something we want to add at this time. Even adding this behind an option flag would possible seem to be cause more issues.

I'm going to close this issue then for now and if you ever want to revisit or come with another approach please do.
